tests/library/test_pydantic.py

```diff
@@ -0,0 +1,261 @@
+
+    # Run with the mock model
+    prepared_string = f"<s>{json_string}"
+    lm = models.Mock(prepared_string.encode(), echo=False)
+    lm += grammar
+
+    # Make sure the round trip works
+    round_trip_object = validate_string(lm[CAPTURE_KEY], pydantic_model)
+    assert round_trip_object == target_obj
+
+
+def check_match_failure(
+    bad_obj: Any,
+    failure_byte: bytes,
+    pydantic_model: Union[Type[pydantic.BaseModel], pydantic.TypeAdapter],
+):
+    bad_string = to_compact_json(bad_obj)
+    grammar = gen_pydantic(pydantic_model)
+    with pytest.raises(ParserException) as pe:
+        grammar.match(bad_string, raise_exceptions=True)
+    assert pe.value.current_byte == failure_byte
+
+
+def test_simple_model():
+    class Simple(pydantic.BaseModel):
+        my_string: str
+
+    my_obj = Simple(my_string="some string")
+    generate_and_check(my_obj, Simple)
+
+
+def test_model_with_int_list():
+    class MyModel(pydantic.BaseModel):
+        my_list: List[int] = pydantic.Field(default_factory=list)
+
+    my_obj = MyModel(my_list=[1, 2, 3, 4])
+    generate_and_check(my_obj, MyModel)
+
+
+def test_nested_model():
+    class A(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_a_str")
+
+    class B(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_b_str")
+        my_A: A = pydantic.Field(default_factory=A)
+
+    class C(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_c_str")
+        my_B: B = pydantic.Field(default_factory=B)
+
+    my_obj = C(my_str="some other string!")
+    generate_and_check(my_obj, C)
+
+
+@pytest.mark.parametrize("has_A", [True, False])
+def test_model_with_optional(has_A):
+    class A(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_a_str")
+
+    class B(pydantic.BaseModel):
+        b_str: str = pydantic.Field(default="Some string")
+        my_A: Union[A, None] = pydantic.Field(default=None)
+
+    if has_A:
+        my_obj = B(my_A=A(my_str="a long string or two"))
+    else:
+        my_obj = B(b_str="A long b string")
+
+    generate_and_check(my_obj, B)
+
+
+@pytest.mark.parametrize("target_obj", ["hello", 42, False])
+def test_literal(target_obj):
+    model = pydantic.TypeAdapter(Literal["hello", 42, False])
+    generate_and_check(target_obj, model)
+
+
+class TestTuple:
+
+    @pytest.mark.parametrize("target_obj", [(1,), (1, 2), (1, 2, 3, 4, 5)])
+    def test_variadic(self, target_obj):
+        model = pydantic.TypeAdapter(Tuple[int, ...])
+        generate_and_check(target_obj, model)
+
+    @pytest.mark.xfail(
+        reason="Underlying guidance.json does not yet support sequences with length specifications"
+    )
+    def test_homogeneous(self):
+        model = pydantic.TypeAdapter(Tuple[float, float, float])
+        generate_and_check((3.14, 2.718, 1.41), model)
+
+    @pytest.mark.xfail(
+        reason="Underlying guidance.json does not yet support prefixItems"
```

Comment:
Could you file an issue to track this? And appropriate issues for the other xfails.

Reply:
Sure thing. Although I have a PR I plan to submit later this evening that will make these pass. Depending on the order of which PR is accepted first (if at all :wink:), the resulting xpass will have to be resolved in one or the other

Reply:
Issue tracked [here](https://github.com/guidance-ai/guidance/issues/731)